src/Symfony/Component/DependencyInjection/Loader/FileLoader.php

```diff
@@ -34,4 +48,47 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
 
         parent::__construct($locator);
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function import($resource, $type = null, $ignoreErrors = false, $sourceResource = null)
+    {
+        if (strlen($resource) === $i = strcspn($resource, '*?{[')) {
+            $directoryPrefix = $resource;
+            $directoryGlob = '';
+        } else {
+            $directoryPrefix = dirname(substr($resource, 0, 1 + $i));
+            $directoryGlob = substr($resource, strlen($directoryPrefix));
+        }
+
+        try {
+            $directoryPrefix = $this->locator->locate($directoryPrefix, $this->currentDir, true);
+            $directoryPrefix = realpath($directoryPrefix) ?: $directoryPrefix;
+            $directoryGlob = $directoryPrefix.$directoryGlob;
+
+            if (!$files = glob($directoryGlob, defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
+                throw new FileLoaderLoadException($directoryGlob, $sourceResource);
+            }
+
+            foreach ($files as $file) {
+                if (is_dir($file)) {
+                    $this->container->addResource(new DirectoryResource($file, '/^$/'));
```

Comment:
This misses tracking new/removed files in $directoryPrefix.
I think it's better to track at the prefix level as done at https://github.com/symfony/symfony/pull/21289/files#diff-ad1ed76aba6a80df5a48dfa4585adcf3R74:
`$this->container->addResource(new DirectoryResource($directoryPrefix, '/^$/'));`